[DoDeca] IGC Task

klshuster opened the pull request:
**Patch description**
This PR mainly adds the [Image Grounded Conversations (IGC)](https://www.aclweb.org/anthology/I17-1047/) task to ParlAI; IGC is one of the 12 [dodecaDialogue](https://parl.ai/projects/dodecadialogue/) tasks.

The IGC task must be manually downloaded from a url and placed in the correct folder; this teacher will automatically download the available images.

I additionally made some changes to the `AbstractImageTeacher`

**Testing steps**
Added `tests/tasks/test_igc.py`. Should only be run locally.

**Logs**
```
$ python tests/tasks/test_igc.py
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216, got 192
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216, got 192
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216, got 192
  return f(*args, **kwds)
/private/home/kshuster/miniconda3/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
..
----------------------------------------------------------------------
Ran 2 tests in 23.807s

OK
```

**Data tests (if applicable)**
```
$ python parlai/scripts/verify_data.py -t igc
...
274s elapsed: {'exs': 4353, '%done': '100.00%', 'time_left': '0s', 'missing_text': 0, 'missing_labels': 0, 'missing_label_candidates': 4353, 'empty_string_label_candidates': 0, 'label_candidates_with_missing_label': 0, 'did_not_return_message': 0}
```

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+            raise RuntimeError(
+                'Please download the IGC Dataset from '
+                'https://www.microsoft.com/en-us/download/details.aspx?id=55324. '
+                'Then, make sure to put the two .csv files in ParlAI/data/igc/'
```
can you sub in here the the actual datapath? the user might supply --datapath

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+
+        return action
+
+    def load_data(self, data_path: str, opt: Dict[str, Any]) -> List[Dict[str, Any]]:
```
opt is an object, `parlai.core.opt.Opt`

stephenroller replied:
thanks for typing :D

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+    """
+
+    def __init__(
+        self: AbstractImageTeacher, opt: Dict[str, Any], shared: Dict[str, Any] = None
```
see comment about opt below

you proposed setting a typing helper module. `Shared` is a really good choice for that, would love if you did that here.

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+
+class IGCTeacher(AbstractImageTeacher):
+    """
+    Teacher for IGC Task.
```
nit: can you expand out the acronym here

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+
+        for multi-reference labels.
+        """
+        AbstractImageTeacher.add_cmdline_args(argparser)
```
do `super().add_cmdline_args(argparser)` instead please

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+            image_id = self.blank_image_id
+        return os.path.join(self.get_image_path(self.opt), image_id)
+
+    def get_data_path(self, opt: Dict[str, Any]) -> str:
```
another opt

stephenroller replied:
also, should this one be private?

klshuster replied:
this function, along with several others, are public methods in the parent `AbstractImageTeacher` class - normally i'd say yes, probably should make private, however these methods are encouraged to be overridden, so not sure what the correct approach is

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+            os.makedirs(image_features_path)
+
+        return os.path.join(
+            image_features_path, '%s_%s_features_dict' % (image_model_name, dt)
```
nit, fstring might look be cool

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+
+        if dt == 'train':
+            # Take first 90% of valid set as train
+            self.data = self.data[: int(len(self.data) * 0.9)]
```
is the data pre-shuffled?

klshuster replied:
to my knowledge no, however there doesn't appear to be any sense of order in the datafile. this is an arbitrary choice anyway, as there is no train set really...

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        IGCTeacher.add_cmdline_args(argparser)
```
similarly, super()

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+        return shared
+
+
+class IGCOneSideTeacher(ABC, IGCTeacher):
```
why ABC here?

klshuster replied:
i don't want anyone instantiating this teacher directly, as one should instead use a subclass (response only, question only)

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,336 @@
+            help='specify true to evaluate on multi-reference labels',
+        )
+
+    def num_episodes(self) -> int:
```
so many types, i'm so happy.

stephenroller commented on tests/tasks/test_igc.py:
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
```
no way!

stephenroller commented on tests/tasks/test_igc.py:
```diff
@@ -0,0 +1,78 @@
+    Basic tests for IGC Teacher to ensure it works for every given variant.
+    """
+
+    def run_display_output(
```
make private?

klshuster commented:
I've addressed your comments @stephenroller. I added a prototype `typing.py` file to `parlai/utils/` --> I wasn't sure exactly what sort of format we wanted to go with here, so please let me know what you think

klshuster commented:
In the typing file, I've adopted a convention where we define a class for each "type", and then define a type for this class. E.g., shared is now defined as `_Shared` with type `TShared`. If this is not desirable, please let me know.

stephenroller commented on parlai/tasks/igc/agents.py:
```diff
@@ -0,0 +1,339 @@
+from parlai.core.params import Opt
+from parlai.core.teachers import AbstractImageTeacher
+
+import parlai.utils.typing as PT
```
I would mind mirroring the regular typing. `From parlai.utils.typing import TShared`, but this is cool too. I'm happy.

stephenroller commented on parlai/utils/typing.py:
```diff
@@ -0,0 +1,24 @@
+from typing import TypeVar
+
+
+class _Shared(dict):
```
huh, I thought somehow it would be like `TShared(Dict[str, Any])`?

stephenroller commented:
note the current output of mypy:

```
parlai/tasks/igc/agents.py:42: error: Argument 2 for "super" not an instance of argument 1
parlai/tasks/igc/agents.py:46: error: "AbstractImageTeacher" has no attribute "blank_image_id"
Found 2 errors in 1 file (checked 786 source files)
```

klshuster commented:
it's not immediately clear to me how to fix those so i am just removing the `self: AbstractImageTeacher` typing from init for now...
